timemachine/maps/estimators.py

```diff
@@ -0,0 +1,56 @@
+"""Utilities for computing mapped works or mapped u_kn matrix, incorporating invertible map_fxns"""
+
+import numpy as np
+from pymbar.utils import kln_to_kn  # TODO: should this be "vendored"?
+
+__all__ = ["compute_mapped_reduced_work", "compute_mapped_u_kn"]
+
+
+def compute_mapped_reduced_work(src_samples, src_reduced_energy_fxn, dst_reduced_energy_fxn, map_fxn):
+    """Compute minus log importance weights, including a change of variables defined by map_fxn
+
+    References
+    ----------
+    [Jarzynski, 2002] Targeted free energy perturbation
+        https://journals.aps.org/pre/abstract/10.1103/PhysRevE.65.046122
+    """
+
+    # compare with version without map_fxn
+    # return dst_reduced_energy_fxn(src_samples) - src_reduced_energy_fxn(src_samples)
+
+    mapped_samples, logdetjacs = map_fxn(src_samples)
+    return dst_reduced_energy_fxn(mapped_samples) - src_reduced_energy_fxn(src_samples) - logdetjacs
+
+
+def compute_mapped_u_kn(sample_lists, reduced_energy_fxns, map_fxns):
```

Comment:
Signature: assumes lists / arrays of callables -- rather than lists / arrays of parameters + a single u_fxn, map_fxn that accepts parameters -- which deviates from idiomatic jax and rules out some jax transformations. Each callable `map_fxn[i, j]` is assumed to act on whole trajectories (possibly including manual or automatic optimizations for processing whole trajectories).

This seemed the most concise and general way to write this utility (e.g. it allows independent implementations for each pair of states -- maybe you want to use the identity fxn for `map_fxns[i, i]`, a simple implementation for `map_fxns[i, j]` where `|j - i| == 1`, a more complicated map for `map_fxns[i, j]` where `|j - i| == 2`, and to not even bother for distant pairs `map_fxns[i, j]` where `|j - i| > 2` or something). However, this may need to be modified for performance if K^2 Python loop overhead is large, compared with the cost of processing each individual trajectory, and if a Jax version of this loop can be made significantly faster.

The current implementation of this utility is a good fit for the terminal bond map implementation, where a reasonably fast function for processing `(traj, map_params)` can be prepared [using `jit/vmap` once](https://github.com/proteneer/timemachine/blob/3602dd0bc89bf20ecdd24a83aee4f8df61496c8e/timemachine/maps/terminal_bonds.py#L100), and then essentially `partial`'d with differing values of `map_params` to [produce](https://github.com/proteneer/timemachine/blob/3602dd0bc89bf20ecdd24a83aee4f8df61496c8e/tests/test_terminal_bond_maps.py#L153-L158) a `K`-by-`K` array of callables (without triggering JIT `K^2` times). I don't anticipate issues with using a similar pattern for other maps in the future.

Reply:
This sounds reasonable to me.

> e.g. it allows independent implementations for each pair of states

Could this also be reasonably achieved using a piecewise function? E.g. it seems relatively straightforward to implement the case of the functional form depending on $|i-j|$ using `jnp.where` or similar.

Reply:
>Could this also be reasonably achieved using a piecewise function? E.g. it seems relatively straightforward to implement the case of the functional form depending on  using jnp.where or similar.

In the first example, yes, but in other foreseeable experimental usages (e.g. if each state is a different molecule, and each pair is augmented with a different set of dummy atoms*), `np.where` mask wouldn't be as readily applicable.

(*Currently unclear if there's a correct way to do this, but it's something that I'll want to try soon.)